core/include/drivers/stm32_rif.h

```diff
@@ -0,0 +1,188 @@
+ */
+
+#ifndef __DRIVERS_STM32_RIF_H
+#define __DRIVERS_STM32_RIF_H
+
+#include <dt-bindings/firewall/stm32mp25-rif.h>
+#include <io.h>
+#include <util.h>
+#include <tee_api_types.h>
+
+/*
+ * CIDCFGR register
+ */
+#define _CIDCFGR_CFEN			BIT(0)
+#define _CIDCFGR_SEMEN			BIT(1)
+#define _CIDCFGR_SEMWL(x)		BIT(SEMWL_SHIFT + (x))
+
+/*
+ * SEMCR register
+ */
+#define _SEMCR_MUTEX			BIT(0)
+#define _SEMCR_SEMCID_SHIFT		U(4)
+#define _SEMCR_SEMCID_MASK		GENMASK_32(6, 4)
+
+/*
+ * Miscellaneous
+ */
+#define MAX_CID_BITFIELD		U(3)
+#define MAX_CID_SUPPORTED		U(7)
+
+#define SCID_SHIFT			U(4)
+#define SEMWL_SHIFT			U(16)
+#define RIF_ID_SHIFT			U(24)
+
+#define RIF_ID_MASK			GENMASK_32(31, 24)
+#define RIF_CHANNEL_ID(x)		((RIF_ID_MASK & (x)) >> RIF_ID_SHIFT)
+
+#define RIFPROT_SEC			BIT(8)
+#define RIFPROT_PRIV			BIT(9)
+#define RIFPROT_LOCK			BIT(10)
+
+/*
+ * struct rif_conf_data - Structure containing RIF configuration data
+ *
+ * @access_mask: Array of the masks of the registers which will be configured.
+ * @sec_conf: Secure configuration registers.
+ * @priv_conf: Privilege configuration registers.
+ * @cid_confs: CID filtering configuration register value for a peripheral
+ *             resource (e.g: GPIO pins, FMC controllers)
+ * @lock_conf: RIF configuration locking registers
+ *
+ * For a hardware block having 56 channels, there will be 56 cid_confs
+ * registers and 2 sec_conf and priv_conf registers
+ */
+struct rif_conf_data {
+	uint32_t *access_mask;
+	uint32_t *sec_conf;
+	uint32_t *priv_conf;
+	uint32_t *cid_confs;
+	uint32_t *lock_conf;
+};
+
+#ifdef CFG_STM32_RIF
+static inline bool stm32_rif_scid_ok(uint32_t cidcfgr, uint32_t scid_m,
+				     uint32_t cid)
+{
+	return (cidcfgr & scid_m) == SHIFT_U32(cid, SCID_SHIFT) &&
+	       !(cidcfgr & _CIDCFGR_SEMEN);
+}
+
+static inline bool stm32_rif_semaphore_enabled_and_ok(uint32_t cidcfgr,
+						      uint32_t wl)
+{
+	return (cidcfgr & _CIDCFGR_CFEN) && (cidcfgr & _CIDCFGR_SEMEN) &&
+	       (cidcfgr & _CIDCFGR_SEMWL(wl));
+}
+
+static inline bool stm32_rif_semaphore_mode_incorrect(uint32_t cidcfgr)
+{
+	return !(cidcfgr & _CIDCFGR_CFEN) || !(cidcfgr & _CIDCFGR_SEMEN) ||
+	       (!(cidcfgr & _CIDCFGR_SEMWL(RIF_CID1)) &&
+		(cidcfgr & _CIDCFGR_SEMEN));
+}
+
+/*
+ * stm32_rif_check_access() - Check every possible configuration where accessing
+ *			      RIF is possible:
+ * -When no CID filtering is enabled on a controller
+ * -When CID filtering is enabled on a controller, semaphore
+ *  is disabled, and static CID corresponds to Cortex-A
+ * -When CID filtering is enabled on a controller, semaphore
+ *  is enabled and is not taken, and Cortex-A CID is
+ *  authorized
+ * -When CID filtering is enabled on a controller, semaphore
+ *  is enabled and is taken, and Cortex-A CID is
+ *  authorized and semaphore is already taken by Cortex-A
+ *
+ * @cidcfgr: CIDCFGR configuration register value
+ * @semcr: SEMCR register value
+ * @nb_cid_supp: Number of CID supported
```

Comment:
for consistency with description of the other functions:
`@nb_cid_supp: Number of supported CID for the peripheral`